### ISSUE TITLE
Lexically declared names should cause ReferenceErrors when used before initialization

### ISSUE THREAD

**Issue by macmv** [boa-dev/boa#1393]
Bug: variable declarations at runtime are not quite correct.

```js
var a = 5;
{
  console.log(a);
  let a = 3;
  console.log(a);
}
```

This prints out `5` then `3`. However, this is invalid javascript. Because `a` was redeclared within the inner block, the first `console.log(a)` should produce an error like this: `Uncaught ReferenceError: Cannot access 'a' before initialization`. I found this issue while implementing variable scoping at parse time (see #1391).

I think this should be checked at parse time, but I'm not entirely sure. If this is checked at parse time, then I would add it to my current PR (again #1391), and use the new `env` variable I've been working on. If this should be checked at runtime, then I would add some checks at the start of executing `StatementList`. There would need to be a way to find all the lexically declared names of a node, but that is not very difficult.

I am free to fix this, but I would like some advice on how/where this should be fixed.

**Comment by jedel1043** [comment]
```Javascript
var a = 5;
{
    console.log("at runtime");
    console.log(a);
    let a = 3;
    console.log(a);
}
```
Running this in V8 outputs:
```
at runtime
VM457:4 Uncaught ReferenceError: Cannot access 'a' before initialization
    at <anonymous>:4:17
```
so we should probably catch it at runtime.

**Comment by jasonwilliams** [comment]
Good find, looks like we need to hoist the declaration binding of those lets/const within the block but leave the initialisation where it is. We do something similar for var so may be worth finding how that works and learning from that.

As @jedel1043 said referenceErrors happen at runtime. Best to look at the environments

**Comment by Razican** [comment]
Actually, hoisting is not spec compliant on Boa. We just move all declarations to the top of the scope, which is not spec compliant, but works for most cases. We should change this, but for that, we will need to have information on declared names at the scope level, and change the parser.

**Comment by jedel1043** [comment]
We would need to implement [`Reference Records`](https://tc39.es/ecma262/#sec-reference-record-specification-type) to solve this, I think.

**Comment by raskad** [comment]
I think this was fixed with #1829. The following code now works as expected:

```js
var a = 5;
{
    console.log("at runtime");
    console.log(a);
    let a = 3;
    console.log(a);
}
```

Node:
```txt
at runtime
ReferenceError: Cannot access 'a' before initialization
```

boa:
```txt
at runtime
Uncaught "ReferenceError": "a is not initialized"
```
